File: mobile/iosApp/Bissbilanz/API/OpenFoodFactsClient.swift
import Foundation

/// Direct Open Food Facts v2 client used in Local mode, where there is no
/// backend to proxy barcode lookups through. Replicates the field selection
/// and mapping of the server proxy (`src/lib/server/openfoodfacts.ts`) and the
/// Android shared `OpenFoodFactsClient`, returning the `Food` prefill shape
/// the barcode scanner uses.
struct OpenFoodFactsClient {
    private let session: URLSession

    private static let apiBase = "https://world.openfoodfacts.org/api/v2/product"
    private static let searchBase = "https://world.openfoodfacts.org/cgi/search.pl"
    private static let userAgent = "Bissbilanz/1.0 (https://bissbilanz.orellbuehler.ch)"
    private static let fields = [
        "product_name", "brands", "nutriscore_grade", "nova_group",
        "additives_tags", "ingredients_text", "image_front_url", "nutriments",
    ].joined(separator: ",")

    private static let gToMg = 1000.0
    private static let gToUg = 1_000_000.0

    /// (Food key, OFF nutriments key, unit conversion) for the optional
    /// extended nutrients — mirrors the Android client's mapping.
    private static let extendedNutrients: [(String, String, Double)] = [
        ("saturatedFat", "saturated-fat_100g", 1), ("monounsaturatedFat", "monounsaturated-fat_100g", 1),
        ("polyunsaturatedFat", "polyunsaturated-fat_100g", 1), ("transFat", "trans-fat_100g", 1),
        ("cholesterol", "cholesterol_100g", gToMg), ("omega3", "omega-3-fat_100g", 1),
        ("omega6", "omega-6-fat_100g", 1), ("sugar", "sugars_100g", 1),
        ("addedSugars", "added-sugars_100g", 1), ("sugarAlcohols", "sugar-alcohols_100g", 1),
        ("starch", "starch_100g", 1), ("sodium", "sodium_100g", gToMg),
        ("potassium", "potassium_100g", gToMg), ("calcium", "calcium_100g", gToMg),
        ("iron", "iron_100g", gToMg), ("magnesium", "magnesium_100g", gToMg),
        ("phosphorus", "phosphorus_100g", gToMg), ("zinc", "zinc_100g", gToMg),
        ("copper", "copper_100g", gToMg), ("manganese", "manganese_100g", gToMg),
        ("selenium", "selenium_100g", gToUg), ("iodine", "iodine_100g", gToUg),
        ("fluoride", "fluoride_100g", gToMg), ("chromium", "chromium_100g", gToUg),
        ("molybdenum", "molybdenum_100g", gToUg), ("chloride", "chloride_100g", gToMg),
        ("vitaminA", "vitamin-a_100g", gToUg), ("vitaminC", "vitamin-c_100g", gToMg),
        ("vitaminD", "vitamin-d_100g", gToUg), ("vitaminE", "vitamin-e_100g", gToMg),
        ("vitaminK", "vitamin-k_100g", gToUg), ("vitaminB1", "vitamin-b1_100g", gToMg),
        ("vitaminB2", "vitamin-b2_100g", gToMg), ("vitaminB3", "vitamin-b3_100g", gToMg),
        ("vitaminB5", "vitamin-b5_100g", gToMg), ("vitaminB6", "vitamin-b6_100g", gToMg),
        ("vitaminB7", "vitamin-b7_100g", gToUg), ("vitaminB9", "vitamin-b9_100g", gToUg),
        ("vitaminB12", "vitamin-b12_100g", gToUg), ("caffeine", "caffeine_100g", gToMg),
        ("alcohol", "alcohol_100g", 1), ("water", "water_100g", 1), ("salt", "salt_100g", 1),
    ]

    init(session: URLSession = .shared) {
        self.session = session
    }

    /// Looks the barcode up on Open Food Facts. Returns nil for unknown
    /// products or unparseable responses; throws on transport errors.
    func lookupBarcode(_ barcode: String) async throws -> Food? {
        // Untrusted scanner input — see the twin in `BissbilanzAPI.lookupBarcode`.
        guard let encoded = barcode.addingPercentEncoding(withAllowedCharacters: .alphanumerics),
              var components = URLComponents(string: "\(Self.apiBase)/\(encoded).json")
        else { return nil }
        components.queryItems = [URLQueryItem(name: "fields", value: Self.fields)]
        guard let url = components.url else { return nil }
        var request = URLRequest(url: url)
        request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200 ..< 300).contains(http.statusCode) else {
            return nil
        }
        guard let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              Self.number(root["status"]) == 1,
              let product = root["product"] as? [String: Any]
        else {
            return nil
        }
        return Self.mapProduct(product, barcode: barcode)
    }

    /// Free-text product search (`/cgi/search.pl`), mirroring the server
    /// proxy's `searchProducts`. Only products with a name and a barcode are
    /// returned so every hit can be instantiated by barcode later.
    func searchProducts(query: String, limit: Int = 10) async throws -> [Food] {
        guard var components = URLComponents(string: Self.searchBase) else { return [] }
        components.queryItems = [
            URLQueryItem(name: "search_terms", value: query),
            URLQueryItem(name: "search_simple", value: "1"),
            URLQueryItem(name: "action", value: "process"),
            URLQueryItem(name: "json", value: "1"),
            URLQueryItem(name: "page_size", value: String(min(max(limit, 1), 20))),
            URLQueryItem(name: "fields", value: "code,\(Self.fields)"),
        ]
        guard let url = components.url else { return [] }
        var request = URLRequest(url: url)
        request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200 ..< 300).contains(http.statusCode) else {
            return []
        }
        guard let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              let products = root["products"] as? [[String: Any]]
        else {
            return []
        }
        return products.compactMap { product in
            guard let code = product["code"] as? String, !code.isEmpty else { return nil }
            guard let food = Self.mapProduct(product, barcode: code), !food.name.isEmpty else { return nil }
            return food
        }
    }

    /// Maps an OFF product onto the `Food` prefill shape (per-100g values,
    /// like the server proxy). Internal for tests.
    static func mapProduct(_ product: [String: Any], barcode: String) -> Food? {
        let nutriments = product["nutriments"] as? [String: Any] ?? [:]

        var dict: [String: Any] = [
            // The proxy has no separate id; the barcode identifies the product.
            "id": barcode,
            "userId": "",
            "name": (product["product_name"] as? String) ?? "",
            "servingSize": 100,
            "servingUnit": "g",
            "calories": number(nutriments["energy-kcal_100g"]) ?? 0,
            "protein": number(nutriments["proteins_100g"]) ?? 0,
            "carbs": number(nutriments["carbohydrates_100g"]) ?? 0,
            "fat": number(nutriments["fat_100g"]) ?? 0,
            "fiber": number(nutriments["fiber_100g"]) ?? 0,
            "barcode": barcode,
            "isFavorite": false,
        ]
        if let brand = product["brands"] as? String, !brand.isEmpty {
            dict["brand"] = brand
        }
        if let grade = product["nutriscore_grade"] as? String,
           ["a", "b", "c", "d", "e"].contains(grade)
        {
            dict["nutriScore"] = grade
        }
        if let novaGroup = number(product["nova_group"]), (1.0 ... 4.0).contains(novaGroup) {
            dict["novaGroup"] = Int(novaGroup)
        }
        if let additives = product["additives_tags"] as? [String] {
            dict["additives"] = additives
        }
        if let ingredientsText = product["ingredients_text"] as? String {
            dict["ingredientsText"] = ingredientsText
        }
        if let imageUrl = product["image_front_url"] as? String {
            dict["imageUrl"] = imageUrl
        }
        for (foodKey, offKey, conversion) in extendedNutrients {
            if let value = extractNutrient(nutriments[offKey], conversion: conversion) {
                dict[foodKey] = value
            }
        }
        return try? JSONPatch.decode(Food.self, from: dict)
    }

    /// Mirrors the proxy's `extractNutrient`: optional unit conversion and
    /// rounding to 2 decimal places, nil when missing or not numeric.
    private static func extractNutrient(_ value: Any?, conversion: Double) -> Double? {
        guard let raw = number(value) else { return nil }
        return (raw * conversion * 100).rounded() / 100
    }

    /// OFF nutriment values may be numbers or numeric strings.
    private static func number(_ value: Any?) -> Double? {
        if let number = value as? NSNumber {
            return number.doubleValue.isNaN ? nil : number.doubleValue
        }
        if let string = value as? String {
            return Double(string)
        }
        return nil
    }
}

File: mobile/shared/src/commonMain/kotlin/com/bissbilanz/api/OpenFoodFactsClient.kt
package com.bissbilanz.api

import com.bissbilanz.api.generated.model.OpenFoodFactsProduct
import com.bissbilanz.createHttpEngine
import io.ktor.client.*
import io.ktor.client.engine.*
import io.ktor.client.plugins.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlin.math.round

/**
 * Direct Open Food Facts v2 client used in Local mode, where the app has no backend
 * to proxy through. Replicates the field selection and mapping of the SvelteKit
 * server proxy (`src/lib/server/openfoodfacts.ts`) so the result matches the
 * generated [OpenFoodFactsProduct] model returned by `/api/openfoodfacts/{barcode}`.
 */
class OpenFoodFactsClient(
    engine: HttpClientEngine = createHttpEngine(),
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    private val client =
        HttpClient(engine) {
            install(HttpTimeout) {
                requestTimeoutMillis = 30_000
                connectTimeoutMillis = 10_000
            }
        }

    suspend fun fetchProduct(barcode: String): OpenFoodFactsProduct? {
        val response =
            client.get("$API_BASE/$barcode.json") {
                parameter("fields", FIELDS)
                header(HttpHeaders.UserAgent, USER_AGENT)
                accept(ContentType.Application.Json)
            }
        if (!response.status.isSuccess()) return null

        val root =
            try {
                json.parseToJsonElement(response.bodyAsText()).jsonObject
            } catch (e: Exception) {
                if (e is kotlinx.coroutines.CancellationException) throw e
                return null
            }
        if ((root["status"] as? JsonPrimitive)?.intOrNull != 1) return null
        val product = root["product"] as? JsonObject ?: return null
        return mapProduct(product, barcode)
    }

    /**
     * Free-text product search, mirroring the server proxy's `searchProducts`
     * (`/cgi/search.pl`). Only products with a name and a barcode are returned so
     * every hit can be instantiated by barcode later.
     */
    suspend fun searchProducts(
        query: String,
        limit: Int = 10,
    ): List<OpenFoodFactsProduct> {
        val response =
            client.get(SEARCH_BASE) {
                parameter("search_terms", query)
                parameter("search_simple", "1")
                parameter("action", "process")
                parameter("json", "1")
                parameter("page_size", limit.coerceIn(1, 20))
                parameter("fields", "code,$FIELDS")
                header(HttpHeaders.UserAgent, USER_AGENT)
                accept(ContentType.Application.Json)
            }
        if (!response.status.isSuccess()) return emptyList()

        val root =
            try {
                json.parseToJsonElement(response.bodyAsText()).jsonObject
            } catch (e: Exception) {
                if (e is kotlinx.coroutines.CancellationException) throw e
                return emptyList()
            }
        val products = root["products"] as? JsonArray ?: return emptyList()
        return products.mapNotNull { element ->
            val product = element as? JsonObject ?: return@mapNotNull null
            val code = product.stringOrNull("code")?.takeIf { it.isNotBlank() } ?: return@mapNotNull null
            mapProduct(product, code).takeIf { it.name.isNotBlank() }
        }
    }

    private fun mapProduct(
        product: JsonObject,
        barcode: String,
    ): OpenFoodFactsProduct {
        val nutriments = product["nutriments"] as? JsonObject ?: JsonObject(emptyMap())

        return OpenFoodFactsProduct(
            // The server proxy has no separate id; the barcode identifies the product.
            id = barcode,
            name = product.stringOrNull("product_name") ?: "",
            brand = product.stringOrNull("brands") ?: "",
            barcode = barcode,
            imageUrl = product.stringOrNull("image_front_url"),
            nutriScore =
                product.stringOrNull("nutriscore_grade")?.let { grade ->
                    OpenFoodFactsProduct.NutriScore.entries.firstOrNull { it.value == grade }
                },
            novaGroup = nutrimentNumber(product, "nova_group")?.takeIf { it in 1.0..4.0 },
            servingSize = 100.0,
            servingUnit = "g",
            calories = nutrimentNumber(nutriments, "energy-kcal_100g") ?: 0.0,
            protein = nutrimentNumber(nutriments, "proteins_100g") ?: 0.0,
            carbs = nutrimentNumber(nutriments, "carbohydrates_100g") ?: 0.0,
            fat = nutrimentNumber(nutriments, "fat_100g") ?: 0.0,
            fiber = nutrimentNumber(nutriments, "fiber_100g") ?: 0.0,
            additives =
                (product["additives_tags"] as? JsonArray)
                    ?.mapNotNull { (it as? JsonPrimitive)?.takeIf { p -> p.isString }?.content }
                    ?: emptyList(),
            ingredientsText = product.stringOrNull("ingredients_text"),
            saturatedFat = extractNutrient(nutriments, "saturated-fat_100g"),
            monounsaturatedFat = extractNutrient(nutriments, "monounsaturated-fat_100g"),
            polyunsaturatedFat = extractNutrient(nutriments, "polyunsaturated-fat_100g"),
            transFat = extractNutrient(nutriments, "trans-fat_100g"),
            cholesterol = extractNutrient(nutriments, "cholesterol_100g", G_TO_MG),
            omega3 = extractNutrient(nutriments, "omega-3-fat_100g"),
            omega6 = extractNutrient(nutriments, "omega-6-fat_100g"),
            sugar = extractNutrient(nutriments, "sugars_100g"),
            addedSugars = extractNutrient(nutriments, "added-sugars_100g"),
            sugarAlcohols = extractNutrient(nutriments, "sugar-alcohols_100g"),
            starch = extractNutrient(nutriments, "starch_100g"),
            sodium = extractNutrient(nutriments, "sodium_100g", G_TO_MG),
            potassium = extractNutrient(nutriments, "potassium_100g", G_TO_MG),
            calcium = extractNutrient(nutriments, "calcium_100g", G_TO_MG),
            iron = extractNutrient(nutriments, "iron_100g", G_TO_MG),
            magnesium = extractNutrient(nutriments, "magnesium_100g", G_TO_MG),
            phosphorus = extractNutrient(nutriments, "phosphorus_100g", G_TO_MG),
            zinc = extractNutrient(nutriments, "zinc_100g", G_TO_MG),
            copper = extractNutrient(nutriments, "copper_100g", G_TO_MG),
            manganese = extractNutrient(nutriments, "manganese_100g", G_TO_MG),
            selenium = extractNutrient(nutriments, "selenium_100g", G_TO_UG),
            iodine = extractNutrient(nutriments, "iodine_100g", G_TO_UG),
            fluoride = extractNutrient(nutriments, "fluoride_100g", G_TO_MG),
            chromium = extractNutrient(nutriments, "chromium_100g", G_TO_UG),
            molybdenum = extractNutrient(nutriments, "molybdenum_100g", G_TO_UG),
            chloride = extractNutrient(nutriments, "chloride_100g", G_TO_MG),
            vitaminA = extractNutrient(nutriments, "vitamin-a_100g", G_TO_UG),
            vitaminC = extractNutrient(nutriments, "vitamin-c_100g", G_TO_MG),
            vitaminD = extractNutrient(nutriments, "vitamin-d_100g", G_TO_UG),
            vitaminE = extractNutrient(nutriments, "vitamin-e_100g", G_TO_MG),
            vitaminK = extractNutrient(nutriments, "vitamin-k_100g", G_TO_UG),
            vitaminB1 = extractNutrient(nutriments, "vitamin-b1_100g", G_TO_MG),
            vitaminB2 = extractNutrient(nutriments, "vitamin-b2_100g", G_TO_MG),
            vitaminB3 = extractNutrient(nutriments, "vitamin-b3_100g", G_TO_MG),
            vitaminB5 = extractNutrient(nutriments, "vitamin-b5_100g", G_TO_MG),
            vitaminB6 = extractNutrient(nutriments, "vitamin-b6_100g", G_TO_MG),
            vitaminB7 = extractNutrient(nutriments, "vitamin-b7_100g", G_TO_UG),
            vitaminB9 = extractNutrient(nutriments, "vitamin-b9_100g", G_TO_UG),
            vitaminB12 = extractNutrient(nutriments, "vitamin-b12_100g", G_TO_UG),
            caffeine = extractNutrient(nutriments, "caffeine_100g", G_TO_MG),
            alcohol = extractNutrient(nutriments, "alcohol_100g"),
            water = extractNutrient(nutriments, "water_100g"),
            salt = extractNutrient(nutriments, "salt_100g"),
        )
    }

    private fun JsonObject.stringOrNull(key: String): String? = (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

    /** OFF nutriment values may be numbers or numeric strings. */
    private fun nutrimentNumber(
        obj: JsonObject,
        key: String,
    ): Double? = (obj[key] as? JsonPrimitive)?.doubleOrNull?.takeIf { !it.isNaN() }

    /**
     * Mirrors the server proxy's `extractNutrient`: optional unit conversion and
     * rounding to 2 decimal places, null when missing or not numeric.
     */
    private fun extractNutrient(
        nutriments: JsonObject,
        offKey: String,
        conversion: Double = 1.0,
    ): Double? = nutrimentNumber(nutriments, offKey)?.let { round(it * conversion * 100) / 100 }

    fun close() {
        client.close()
    }

    companion object {
        private const val API_BASE = "https://world.openfoodfacts.org/api/v2/product"
        private const val SEARCH_BASE = "https://world.openfoodfacts.org/cgi/search.pl"
        private const val USER_AGENT = "Bissbilanz/1.0 (https://bissbilanz.orellbuehler.ch)"
        private const val FIELDS =
            "product_name,brands,nutriscore_grade,nova_group,additives_tags," +
                "ingredients_text,image_front_url,nutriments"
        private const val G_TO_MG = 1_000.0
        private const val G_TO_UG = 1_000_000.0
    }
}
